contracts/coconsensus/Coconsensus.sol

```diff
@@ -46,23 +47,230 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+        auxiliaryMetachainId = genesisAuxiliaryMetachainId;
+
+        /*
+         * Setup self protocore contract first. All other protocores will
+         * require the dynasty from the self protocore to finalise the genesis
+         * checkpoint.
+         */
+        setupProtocores(auxiliaryMetachainId);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+            /*
+             * The setup of self protocore is already done as a first step so
+             * skip the protocore setup if the current metachain id is equal to
+             * the auxiliary metachain id.
+             */
+            if (currentMetachainId != auxiliaryMetachainId) {
+                // Setup protocore contract for the given metachain id.
+                setupProtocores(currentMetachainId);
+            }
+
+            // Setup observer contract for the given metachain id.
+            setupObservers(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Anchor the state root in to the observer contracts.
+     *
+     * @param _metachainId Metachain id.
+     * @param _rlpBlockHeader RLP encoded block header.
+     *
+     * /pre `_metachainId` is not 0.
+     * /pre `_rlpBlockHeader` is not 0.
+     * /pre `blockHash` should exist in blockchains.
+     * /pre The dynasty of the reported block should be less than current dynasty.
+     * /pre The reported block should be finalized.
+     *
+     * /post Anchor the state root in the observer contract.
+     */
+    function observeBlock(
+        bytes32 _metachainId,
+        bytes calldata _rlpBlockHeader
+    )
+        external
+    {
+        require(
+            _metachainId != bytes32(0),
+            "Metachain id must not be null."
+        );
+
+        require(
+            _rlpBlockHeader.length != 0,
+            "RLP block header must not be null."
+        );
+
+        // Decode the rlp encoded block header.
+        BlockHeader.Header memory blockHeader = BlockHeader.decodeHeader(_rlpBlockHeader);
+
+        Block memory blockStatus = blockchains[_metachainId][blockHeader.height];
+
+        require(
+            blockStatus.blockHash == blockHeader.blockHash,
+            "Provided block header is not valid."
+        );
+
+        require(
+            blockStatus.commitStatus > CheckpointCommitStatus.Committed,
+            "Block must be at least finalized."
+        );
+
+        // Get the current dynasty from the self protocore.
+        ProtocoreI protocore = protocores[auxiliaryMetachainId];
+        uint256 currentDynasty = protocore.dynasty();
```

Comment:
I believe, we have agreed (please correct me if I do not remember properly), that dynasty is kept in coconsensus.

Reply:
yes, changed the code to use `relativeSelfDynasty`